### PR TITLE
refactor of attributes

### PR DESCRIPTION
**Pull request by zacharywelch**
`attributes.rb` needs a little routine maintenance and tlc 💅 💅 

- reuse regular expression for predicate and writer methods in `respond_to_missing`
- move `initialize_collection` and `use_setter_methods` to ClassMethods section with other class methods
- rename `initialize_collection` to `instantiate_collection` and add an `instantiate_record` method similar to [ActiveResource](https://github.com/rails/activeresource/blob/c38f18ce4ee6c83e034e7b465459fdd547804b64/lib/active_resource/base.rb#L1061).
- replace duplication of `new_from_parsed_data` and `instantiate_record` from `http` and other files
- apply 80 character limit for readability

### CODE REVIEW

**Comment by shashiprab** [comment]
Any movement on this? The longer it sits around, the more stale it gets.

**Comment by zacharywelch** [comment]
Hi, was trying to defer these changes until we had a chance to release 0.9.0. Now that's taken care of perhaps we'll merge this weekend. Until then feel free to add any comments or give it a review 👍

**Comment by shashiprab** [comment]
Cool. Appreciate your efforts that have moved the needle over the last month tremendously! 

-Shashi

**Comment by shashiprab** [comment]
Zachary

If you could share your email with me I have proposal about the Her gem. My email is shaship@yahoo.com 

Thanks

-Shashi

File: lib/her/model/attributes.rb [hunk 1]
```diff
@@ -177,42 +144,92 @@ def attribute?(attribute)
...
+        end
+
+        # Use setter methods of model for each key / value pair in params
+        # Return key / value pairs for which no setter method was defined on the
+        # model
+        #
+        # @private
+        def use_setter_methods(model, params = {})
+          reserved = [:id, model.class.primary_key, *model.class.association_keys]
+          model.class.attributes *params.keys.reject { |k| reserved.include?(k) }
```
Comment (huangjun604) [inline lib/her/model/attributes.rb]: Hi @zacharywelch , is there any reason to ignore the string keys?

Comment (zacharywelch) [reply lib/her/model/attributes.rb]: Good catch 👍Reserved keys are stored as symbols but attributes can be assigned with string keys. It should instead read 

```ruby
reserved.include?(k.to_sym)
```

Would you like to submit a PR?

Comment (huangjun604) [reply lib/her/model/attributes.rb]: Thanks for response @zacharywelch . Please help to update it, thanks.

**Pull request merged**